### PR TITLE
🐛 [Story performance] Install page-vh on amp-story-page

### PR DESCRIPTION
**Pull request by mszylkowski**
For CLS concerns, we want to install the `dvh` polyfill as soon as possible, so this PR moves the initialization of the dvh polyfill to the buildCallback. 

The `story-page-vh` CSS var is created in `amp-story-page` so that every time the page has a new size, the callback is called and it updates the page size.

### CODE REVIEW

**Review by gmajoulet — COMMENTED** [review]
The dvh polyfill and the PAGE_SIZE are two different things and I don't understand why we're trying to hard to keep them just one method in all our previous PRs. This is what has been causing so much back and forth.

```
buildCallback()
  if needsDvh
    this.installDvhPolyfill();
    this.getViewport().onResize(() => this.installDvhPolyfill());
```

And then there's no reason for the `PAGE_SIZE` property to be set in amp-story.js over amp-story-page.js as it was before. This is the best heuristics we could have, where we get notified on each story-page resize and get the size as soon as it is available. Plz refer to the previous (pre-bug) implementation.

**Comment by mszylkowski** [comment]
The previous implementation is a deprecated API: https://github.com/ampproject/amphtml/pull/37003/files#diff-ccc4b16bc11b77f8f473c989455195f4f9ddba495a64a7db182e54871f62443aR1052

Also this way we can make it so that the `story-page-vh` gets calculated after `story-dvh`, because the story page size depends on the story-dvh variable on mobile.

**Comment by gmajoulet** [comment]
We've been struggling with this piece of code just because it does not make sense to have them coupled, and have them both in this file.
- we set `PAGE_SIZE` in a method called `viewport units`, which it's not
- we have no guarantee that the story-page size is fully set in amp-story.layoutCallback, we're just putting it there because "it works today on our local machine"

File: extensions/amp-story/1.0/amp-story-page.js [hunk 1]
```diff
@@ -528,6 +534,43 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+          return;
+        }
+        this.storeService_.dispatch(Action.SET_PAGE_SIZE, {height, width});
+        if (!this.cssVariablesStyleEl_) {
+          const doc = this.win.document;
+          this.cssVariablesStyleEl_ = doc.createElement('style');
+          this.cssVariablesStyleEl_.setAttribute('type', 'text/css');
+          doc.head.appendChild(this.cssVariablesStyleEl_);
+        }
+        this.cssVariablesStyleEl_.textContent = `:root {--story-page-vh: ${px(
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:567]: Why is this needed? I thought everything, including --story-page-vh, was based on `--story-dvh` now, can you explain why this entire `cssVariablesStyleEl_` is needed?

File: extensions/amp-story/1.0/amp-story-page.js [hunk 2]
```diff
@@ -528,6 +534,43 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+        this.layoutBox_.height === layoutBox.height)
+    ) {
+      return;
+    }
+    this.getVsync().runPromise({
+      measure: (state) => {
+        const {height, width} = layoutBox;
+        state.height = height;
+        state.width = width;
+        state.vh = height / 100;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:553]: This doesn't read from the DOM and doesn't need to be in a `mutate`, + you're already getting the `layoutBox` a few lines above

File: extensions/amp-story/1.0/amp-story-page.js [hunk 3]
```diff
@@ -528,6 +534,43 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+    const layoutBox = this.getLayoutSize();
+    if (
+      !isPrerenderActivePage(this.element) ||
+      (this.layoutBox_ &&
+        this.layoutBox_.width === layoutBox.width &&
+        this.layoutBox_.height === layoutBox.height)
+    ) {
+      return;
+    }
+    this.getVsync().runPromise({
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:548]: Do you need the promise? Why not use `this.mutateElement()`?

**Comment by mszylkowski** [comment]
@gmajoulet I brought back the code from https://github.com/ampproject/amphtml/pull/37003 but was clearly more complex than it should have. I simplified it now so PTAL.

File: extensions/amp-story/1.0/prerender-active-page.js [hunk 4]
```diff
@@ -2,19 +2,26 @@ import {escapeCssSelectorIdent} from '#core/dom/css-selectors';
...
+
 /**
  * Returns true if the page should be prerendered (for being an active page or
  * first page).
  * @param {!AmpElement} pageElement
  * @return {boolean}
  */
 export function isPrerenderActivePage(pageElement) {
+  if (prerenderActivePage != null) {
+    return prerenderActivePage === pageElement;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/prerender-active-page.js:16]: I think it is important for this function to stay pure. As a coding pattern helper functions should really be pure, + storing DOM elements in helpers can super easily lead to memory leaks as it holds hard references and prevents them from being garbage collected.

Can you have either a:
- `isPrerenderActivePage` or something as an `amp-story-page` ivar
- or just store the ID instead of the full page element (I still think it's an anti pattern)

File: extensions/amp-story/1.0/amp-story-page.js [hunk 5]
```diff
@@ -156,7 +156,10 @@ export const NavigationDirection = {
 export class AmpStoryPage extends AMP.BaseElement {
   /** @override @nocollapse */
   static prerenderAllowed(element) {
-    return isPrerenderActivePage(element);
+    if (!AmpStoryPage.prerenderPage) {
+      AmpStoryPage.prerenderPage = getPrerenderActivePage(element);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:160]: I honestly don't know how I feel about a static property storing a DOM element as a pattern

File: extensions/amp-story/1.0/amp-story-page.js [hunk 6]
```diff
@@ -528,6 +531,19 @@ export class AmpStoryPage extends AMP.BaseElement {
     ]);
   }
 
+  /** @override */
+  onLayoutMeasure() {
+    const {height, width} = this.getLayoutSize();
+    if (
+      !AmpStoryPage.prerenderAllowed(this.element) ||
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:538]: You're tying how many times we run the PAGE_SIZE to how many pages are prerendered, which is unrelated.
All we need is making sure this only runs once, which doesn't have anything to do with which page is initially visible or how many are prerendered.

```
AmpStoryPage {
  constructor() {
    this.isFirstPage_ = matches(this.element, 'amp-story-page:first-of-type');
  }
}
```

And you're done without storing DOM elements, tying business logic to internal behavior, or using static properties.

File: extensions/amp-story/1.0/amp-story-grid-layer.js [hunk 7]
```diff
@@ -61,7 +61,7 @@ export let PresetDetails;
 export class AmpStoryGridLayer extends AmpStoryBaseLayer {
   /** @override @nocollapse */
   static prerenderAllowed(element) {
-    return isPrerenderActivePage(element.parentElement);
+    return AmpStoryPage.prerenderAllowed(element.parentElement);
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-grid-layer.js:64]: Unrelated to this PR, plz remove

File: extensions/amp-story/1.0/prerender-active-page.js [hunk 8]
```diff
@@ -3,10 +3,9 @@ import {parseQueryString} from '#core/types/string/url';
 import {getWin} from '#core/window';
 
 /**
- * Returns true if the page should be prerendered (for being an active page or
- * first page).
- * @param {!AmpElement} pageElement
- * @return {boolean}
+ * Retrieves the page that should be the prerender page.
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/prerender-active-page.js:6]: The previous comment was more helpful, it is not necessarily the prerender page but is the active one, regardless of the visibilitystate

**Comment by mszylkowski** [comment]
Storing the element instead of a boolean is more efficient because every boolean for every instance needs to run the querySelector on the whole document. This means that on every grid layer and on every amp-story-page, it's running the querySelector at least once. It's unrelated to this PR though, so I'm ok with not caching it here, just mentioning this as a disadvantage.

**Review by gmajoulet — APPROVED** [review]

**Comment by mszylkowski** [comment]
@jridgewell for owners approvals. We want to keep using `onLayoutMeasure` until it gets removed due to it having better support than `ResizeObserver`.

**Review by jridgewell — APPROVED** [review]
ResizeObserver is polyfilled in all AMP usages, so you could transition easily.

**Comment by mszylkowski** [comment]
Closing to re-trigger CLA
